ubuntu_image/builder.py

```diff
@@ -289,36 +285,45 @@ def prepare_filesystems(self):
         self.images = os.path.join(self.workdir, '.images')
         os.makedirs(self.images)
         # The image for the boot partition.
-        self.boot_images = []
+        self.part_images = []
         volumes = self.gadget.volumes.values()
         assert len(volumes) == 1, 'For now, only one volume is allowed'
         volume = list(volumes)[0]
+        farthest_offset = 0
         for partnum, part in enumerate(volume.structures):
-            # The root file system implicitly lives right after the farthest
-            # out structure, which may not be the last named structure in the
-            # gadget.yaml.
-            self.rootfs_offset = max(
-                self.rootfs_offset, (part.offset + part.size))
             part_img = os.path.join(self.images, 'part{}.img'.format(partnum))
-            self.boot_images.append(part_img)
-            run('dd if=/dev/zero of={} count=0 bs={} seek=1'.format(
-                part_img, part.size))
-            if part.filesystem is FileSystemType.vfat:
-                label_option = (
-                    '-n {}'.format(part.filesystem_label)
-                    # XXX I think this could be None or the empty string, but
-                    # this needs verification.
-                    if part.filesystem_label
-                    else '')
-                # XXX: hard-coding of sector size
-                run('mkfs.vfat -s 1 -S 512 -F 32 {} {}'.format(
-                    label_option, part_img))
-        # Sanity check whether everything will fit on the disk if an explicit
-        # size was given on the command line.  XXX: This hard-codes last 34
-        # 512-byte sectors for backup GPT, empirically derived from sgdisk
-        # behavior.
-        calculated = ceil(
-            (self.rootfs_offset + self.rootfs_size) / 1024 + 17) * 1024
+            if part.role is StructureRole.system_data:
+                # The image for the root partition.
+                if part.size < self.rootfs_size:
+                    _logger.warning('Rootfs partition size ({}) smaller than '
+                                    'actual rootfs contents {}.',
+                                    self.rootfs_size, part.size)
+                    part.size = self.rootfs_size
+                # We defer creating the root file system image because we have
+                # to populate it at the same time. See mkfs.ext4(8) for details
+                with open(part_img,  'w'):
+                    pass
```

Comment:
Ok, we could. When adding new features I tend to just re-use the existing code since this simplifies the diff and makes it easier to see how the parts moved. I can change it anyway.